056-zk-kraft-migration.md

```diff
@@ -0,0 +1,247 @@
+
+### Finite State Machine (FSM)
+
+The migration process can be described with a Finite State Machine (FSM).
+The operator has a new `MigrationManager` component in charge of tracking the migration finite state machine.
+Transitioning across the states depends on the current state, stored in the `Kafka.status.migrationState` field, and the "input" represented by the user interaction with the operator through the following annotations:
+
+* `strimzi.io/kraft-migration`: to be applied on the `Kafka` custom resource with `true|false` values to start/end the overall migration process.
+* `strimzi.io/kraft-migration`: to be applied on the `KafkaNodePool`(s) custom resource with `dual|kraft|zk` values to represent the the desired "direction" in the migration process by the user:
+  * `dual`: allows to move the migration process forward to reach the point of the "dual-write" with KRaft controllers deployed but with metadata written to both ZooKeeper and KRaft. Brokers are also connected to both ZooKeeper nodes and KRaft controllers.
+  * `kraft`: allows to move the migration process forward from the "dual-write" to the end of the migration with ZooKeeper out of the picture.
+  * `zk`: allows to move the migration process back, when still possible, to have brokers working again with ZooKeeper ensemble only.
+
+The state machine is described in the following diagram:
+
+![Migration Finite State Machine](./images/056-zk-kraft-migration-fsm.png)
+
+The following sections explain the states and transitions in details.
+
+#### Phase 1: Provisioning the KRaft controller quorum
+
+Starting from a `NoMigration` (or null) state (as per `Kafka.status.migrationState` field), the user applies the `strimzi.io/kraft-migration: true` on the `Kafka` custom resource to make the operator aware of the migration "intention".
+The FSM moves to a `WaitControllersInMigration` state where the operator expects the user to create the KRaft controllers pool.
+The user creates the `KafkaNodePool` custom resource containing the definition of the KRaft controllers which are going to replace the ZooKeeper nodes.
+These controllers needed to be deployed in "migration mode", so the `KafkaNodePool` custom resource has to be annotate with `strimzi.io/kraft-migration: dual`.
+
+This way, the operator configures the controllers accordingly with the ZooKeeper migration flag and the connection to the ZooKeeper ensemble. 
+
+```shell
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
+```
+
+> NOTE: without an annotation to differentiate the operator behavior, the operator itself just creates the controller nodes in the pool and restarts the Kafka brokers to use them. ZooKeeper is not used anymore but no proper migration takes place (from ZooKeeper to KRaft metadata).
+
+The KRaft controllers should also be provisioned with the same cluster ID as the existing Kafka cluster.
```

Comment:
While this is anyway something mentioned in the Kafka migration guide, it's already enforced and guaranteed by the operator. I will remove this sentence because doesn't add any value here.